core/src/main/scala/org/apache/spark/deploy/history/EventLogFilterRateCalculator.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import org.apache.hadoop.fs.{FileSystem, Path}
+
+import org.apache.spark.scheduler.SparkListenerEvent
+
+/**
+ * This class calculates the rate of events being accepted, via two phases reading:
+ *
+ * 1) Initialize available [[EventFilterBuilder]] instances, and replay the event log files with
+ * builders, so that these builders can gather the information to create [[EventFilter]] instances.
+ * 2) Initialize [[EventFilter]] instances from [[EventFilterBuilder]] instances, and replay the
+ * event log files with filters. Counts the number of events vs accepted events and calculate
+ * the rate.
+ */
+class EventLogFilterRateCalculator(fs: FileSystem) {
+  def calculate(eventLogPaths: Seq[Path]): Double = {
+    val builders = EventFilterBuilder.initializeBuilders(fs, eventLogPaths)
```

Comment:
I was looking at how this class is used later; if I followed correctly, it's called every time there is a new log file, to see if the files can be compacted. Compaction, though, only happens when trying to rebuild the app's UI.

I'm not sure I see the benefit of that. For starters, this method is as expensive as compacting the files, since you're replaying all the log files twice too.

Then, you're gating compaction on whether someone wants to look an an app's UI. So if nobody ever does, then no compaction ever happens.

It would be good to attempt something smarter here; maybe the listeners can get some idea of how much data would be filtered out in the first scan, so that if you don't detect a lot of stuff you avoid the second scan altogether; and if you do the second scan, might as well just write the compacted data, without waiting for a trigger later.

Reply:
> if I followed correctly, it's called every time there is a new log file, to see if the files can be compacted.

No, it's called only once per application log (app + attempt maybe), and we take only one log file to calculate acceptance rate. `LogInfo` and `AttemptInfoWrapper` are changed to store the decision. (Please let me know if you found that the code doesn't do the thing correctly.)

That's why I decided not to do rate calculating and compacting together. There's still redundant effort for reading a log file twice for rate calculating and compacting, but I guess this is OK given the calculation does occur only once.

Btw, is that OK to compact these event logs (would take quite times) when loading app's UI? My first plan was doing it as one of maintenance workload like cleaning code, but as compaction "replaces" some log files with new one, there would be thread-safety issue between cleaning and loading app's UI, which I wanted to avoid unless I got working code except that, and we would really want to.